### PR TITLE
Add symbol scopes.

### LINKED ISSUE #7120: Feature Request: Allow outline panel to show access specifier (i.e. private, public, protected)
I found #3117 and microsoft/vscode#113845 discussing similar problem. Actually I only want some extra small, grey text showing `private` or `public` beside the names of class members, just like what "declaration" does now.

**Screenshots**
Screenshot from 2021-03-05 21-02-02

Displaying other keywords, such as `override`, `virtual` would also be helpful.

_Originally posted by @YalandHong in https://github.com/microsoft/vscode-cpptools/discussions/7112_

### PR DESCRIPTION
**Pull request by sean-mcmanus**
Fixes https://github.com/microsoft/vscode-cpptools/issues/7120

image

Is the "private: ", "protected:" labeling okay? I figure we actually want https://github.com/microsoft/vscode/issues/113845 .

### CODE REVIEW

File: Extension/src/LanguageServer/Providers/documentSymbolProvider.ts [hunk 1]
```diff
@@ -16,7 +16,12 @@ export class DocumentSymbolProvider implements vscode.DocumentSymbolProvider {
         const documentSymbols: vscode.DocumentSymbol[] = [];
         if (symbols) {
             symbols.forEach((symbol) => {
-                const detail: string = util.getLocalizedString(symbol.detail);
+                let detail: string = util.getLocalizedString(symbol.detail);
+                if (symbol.scope === SymbolScope.Private) {
+                    detail = "private" + (detail.length === 0 ? "" : ": ") + detail;
```
Comment (michelleangela) [inline Extension/src/LanguageServer/Providers/documentSymbolProvider.ts]: Do "private" and "protected" need to be localized or are they C++ keywords in this context?

Comment (sean-mcmanus) [reply Extension/src/LanguageServer/Providers/documentSymbolProvider.ts]: They directly refer to C++ keywords so shouldn't be localized, but I'll look into adding a localization for the punctuation.

Comment (sean-mcmanus) [reply Extension/src/LanguageServer/Providers/documentSymbolProvider.ts]: I added a localizable string for the punctuation.

**Review by michelleangela — APPROVED** [review]

File: Extension/src/LanguageServer/Providers/workspaceSymbolProvider.ts [hunk 2]
```diff
@@ -30,9 +30,9 @@ export class WorkspaceSymbolProvider implements vscode.WorkspaceSymbolProvider {
             let name: string = symbol.name;
             if (suffix.length) {
                 if (symbol.scope === SymbolScope.Private) {
-                    suffix = "private" + localize("c.cpp.workspacesymbolscope.separator", ", ") + suffix;
+                    suffix = localize("c.cpp.symbolscope.separator", "{0}, {1}", "private", suffix);
```
Comment (Colengms) [inline Extension/src/LanguageServer/Providers/workspaceSymbolProvider.ts]: You might want to add a 'hint' here for translators, who might not understand what they should do with `{0}, {1}`.

Adding a hint would look something like:

```
suffix = localize({ key: "c.cpp.symbolscope.separator", comment: ["Used in a list as a separator between keywords.  Example: ..."] }, "{0}, {1}", "private", suffix);
```

**Review by Colengms — CHANGES_REQUESTED** [review]

File: Extension/src/LanguageServer/Providers/workspaceSymbolProvider.ts (hunk 2 above)
Comment (sean-mcmanus) [reply Extension/src/LanguageServer/Providers/workspaceSymbolProvider.ts]: Sure, I wasn't sure if "hint" was available or not yet.

File: Extension/src/LanguageServer/Providers/workspaceSymbolProvider.ts [hunk 3]
```diff
@@ -22,10 +26,21 @@ export class WorkspaceSymbolProvider implements vscode.WorkspaceSymbolProvider {
...
         // Convert to vscode.Command array
         symbols.forEach((symbol) => {
-            const suffix: string = util.getLocalizedString(symbol.suffix);
+            let suffix: string = util.getLocalizedString(symbol.suffix);
             let name: string = symbol.name;
             if (suffix.length) {
+                if (symbol.scope === SymbolScope.Private) {
+                    suffix = localize("c.cpp.symbolscope.separator", "{0}, {1}", "private", suffix);
+                } else if (symbol.scope === SymbolScope.Protected) {
+                    suffix = localize("c.cpp.symbolscope.separator", "{0}, {1}", "protected", suffix);
```
Comment (Colengms) [inline Extension/src/LanguageServer/Providers/workspaceSymbolProvider.ts]: Also, could you assign "private" or "protected" to a variable, so only one call to localize() is necessary?  Multiple calls with the same ID should be fine, but I'm not entire certain what happens in that case.

**Review by Colengms — CHANGES_REQUESTED** [review]

File: Extension/src/LanguageServer/Providers/workspaceSymbolProvider.ts (hunk 3 above)
Comment (sean-mcmanus) [reply Extension/src/LanguageServer/Providers/workspaceSymbolProvider.ts]: I created a util function reused by all 4 cases.

File: Extension/src/common.ts [hunk 4]
```diff
@@ -1049,7 +1050,7 @@ export function escapeForSquiggles(s: string): string {
                 newResults += "\\";
             }
             lastWasBackslash = false;
-            lastBackslashWasEscaped  = false;
```
Comment (sean-mcmanus) [inline Extension/src/common.ts:1052]: I ran file formatting.

**Review by Colengms — APPROVED** [review]

**Pull request merged**